src/FastConv/FastConv.cpp

```diff
@@ -12,21 +13,298 @@ CFastConv::~CFastConv( void )
+
+//====================================================================
+CFastConvFreq::CFastConvFreq(float* pfIr, int iLengthOfIr, int iBlockLength) :
+    CFastConvBase(pfIr, iLengthOfIr)
+{
+    // init fft
+    CFft::createInstance(m_pFFT);
+    m_pFFT->initInstance(iBlockLength, 2, CFft::kWindowHann, CFft::kNoWindow);
+
+    m_iBlockLength = m_pFFT->getLength(CFft::kLengthData);
+    m_iFftLength   = m_pFFT->getLength(CFft::kLengthFft);
+    m_iNumIrBlocks = static_cast<int>(m_iLengthOfIr / static_cast<float>(m_iBlockLength)) + 1;
+    
+    // init real/imag buffers
+    m_pfFFTReal = new float[m_iBlockLength + 1];
+    m_pfFFTImag = new float[m_iBlockLength + 1];
+    m_pfFFTRealCurr = new float[m_iBlockLength + 1];
+    m_pfFFTImagCurr = new float[m_iBlockLength + 1];
+
+    m_ppfIRFreqReal = new float* [m_iNumIrBlocks];
+    m_ppfIRFreqImag = new float* [m_iNumIrBlocks];
+    for (int i = 0; i < m_iNumIrBlocks; i++)
+    {
+        // init ir frequency domain buffers
+        m_ppfIRFreqReal[i] = new float[m_iBlockLength + 1];
+        m_ppfIRFreqImag[i] = new float[m_iBlockLength + 1];
+    }
+    
+    // Buffer to use for zero padding and for storing fft
+    m_pfProcessBuf = new float[m_iFftLength];
+
+    // Init tail -- I don't have a good reason for this particular length other than it's what I concluded from inspecting the chart from the provided textbook
+    m_iLengthOfTail = m_iBlockLength * m_iNumIrBlocks + m_iBlockLength;
+    m_pfTail = new float[m_iLengthOfTail];
+    CVectorFloat::setZero(m_pfTail, m_iLengthOfTail);
+    
+    float* cur = m_pfIr;
+    for (int block = 0; block < m_iNumIrBlocks; block++)
+    {
+        // Zero pad ir block
+        CVectorFloat::setZero(m_pfProcessBuf, m_iFftLength);
+        CVectorFloat::copy(m_pfProcessBuf, cur, std::min<int>(m_iBlockLength, m_iLengthOfIr - (block * m_iBlockLength)));
+
+        // Do FFT and scale output
+        m_pFFT->doFft(m_pfProcessBuf, m_pfProcessBuf);
+        CVectorFloat::mulC_I(m_pfProcessBuf, m_iFftLength, m_iFftLength);
+
+        // Split ir spectrum and save
+        m_pFFT->splitRealImag(m_ppfIRFreqReal[block], m_ppfIRFreqImag[block], m_pfProcessBuf);
+
+        // Update which sample 'cur' points to in 'm_pfIr'
+        cur += iBlockLength;
+    }
+    
+}
+
+CFastConvFreq::~CFastConvFreq()
+{
+
+    for (int i = 0; i < m_iNumIrBlocks; i++)
+    {
+        delete[] m_ppfIRFreqReal[i];
+        delete[] m_ppfIRFreqImag[i];
+    }
+    delete[] m_ppfIRFreqReal;
+    delete[] m_ppfIRFreqImag;
+    delete[] m_pfTail;
+    delete[] m_pfFFTReal;
+    delete[] m_pfFFTImag;
+    delete[] m_pfFFTRealCurr;
+    delete[] m_pfFFTImagCurr;
+    delete[] m_pfProcessBuf;
+    CFft::destroyInstance(m_pFFT);
+
+    CFft* m_pFFT = 0;
+    m_pfProcessBuf = 0;
+    m_pfFFTReal = 0;
+    m_pfFFTImag = 0;
+    m_pfFFTRealCurr = 0;
+    m_pfFFTImagCurr = 0;
+    m_ppfIRFreqReal = 0;
+    m_ppfIRFreqImag = 0;
+}
+
```

Comment:
The two class approach like our first assignment is a very clean design for this assignment! Great job